channels/OFF-CHAIN.md

```diff
@@ -1009,3 +1010,56 @@ contract update](../serializations.md#channel-off-chain-update-call-contract). I
 information needed for a contract call, including the contract address. The other participant co-signs
 the changes and the contract call is considered to be executed. Its results can be extracted from the calls tree in the state tree.
 
+### Contracts referring to on-chain data
+
+Contracts being used in channels off-chain calls have the exact same semantics
```

Comment:
I see [the “off-chain updates” are](https://github.com/aeternity/protocol/blob/1c5b17b44da6ecd661bac0fe86c2b1c935f99e8d/serializations.md#channel-off-chain-update): transfer, deposit, withdrawal, create contract, call contract.

I see [all off-chain updates’ **serializations** are part of consensus](https://github.com/aeternity/protocol/blame/1c5b17b44da6ecd661bac0fe86c2b1c935f99e8d/serializations.md#L692-L701).

I see that [the **semantics** of only the off-chain update call contract is part of consensus](https://github.com/aeternity/protocol/blame/1c5b17b44da6ecd661bac0fe86c2b1c935f99e8d/channels/ON-CHAIN.md#L468-L471). As such, its semantics needs to be detailed. I see this semantics is described in two places:

* This paragraph.

  > Contracts being used in channels off-chain calls have the exact same semantics as one being used on-chain. Because of the different environment though, those might have different results as the on-chain ones.

* [Channels ON-CHAIN document](https://github.com/aeternity/protocol/blame/1c5b17b44da6ecd661bac0fe86c2b1c935f99e8d/channels/ON-CHAIN.md#L468-L471).

  > The update MUST be a call to a contract. The caller of this update MUST be the
  > poster of the force progress transaction. `gas` and `gas_price` are
  > specified in the update as well. The gas fees are going to be paid by the poster
  > of the transaction.

Recommendations:

* I suggest to amend this paragraph with the link to the referred semantics, i.e.

  > Contracts being used in channels off-chain calls have the exact same semantics as [one being used on-chain](/contracts/contract_transactions.md#contract-call-transaction). Because of the different environment though, those might have different results as the on-chain ones.
* The [section in the ON-CHAIN document](https://github.com/aeternity/protocol/blame/1c5b17b44da6ecd661bac0fe86c2b1c935f99e8d/channels/ON-CHAIN.md#L468-L471) shall be integrated mentioning that [amount](https://github.com/aeternity/protocol/blame/1c5b17b44da6ecd661bac0fe86c2b1c935f99e8d/serializations.md#L681) is specified in the update as well, and what the semantics of this movement of coins is (there are off-chain on-chain trees, plus only the accounts of the initiator and responder “make it” as side effect of the force progress - plus state_hash/round of course).

Reply:
Re semantics of the off-chain upadte call contract (part of consensus) it seems it would be beneficial to document:
* Semantics re [gas](https://github.com/aeternity/epoch/pull/1592#discussion_r222615104)
* Semantics re [amount](https://github.com/aeternity/epoch/pull/1592#discussion_r222622621)

Reply:
And it looks like the charging of the amount depends also on the channel reserve in the channel object in the state tree.